crates/bevy_ecs/src/schedule/executor/single_threaded.rs

```diff
@@ -44,13 +61,57 @@ impl SystemExecutor for SingleThreadedExecutor {
     }
 
     fn run(&mut self, schedule: &mut SystemSchedule, world: &mut World) {
+        /// return the next system not exempt from stepping starting at, and
+        /// inclusive of, first.  If there are no more systems that can be
+        /// stepped, the first non-exempt system index will be returned.
+        fn next_system(schedule: &SystemSchedule, first: usize) -> usize {
+            for i in first..schedule.systems_with_stepping.len() {
```

Comment:
This expensive check should only be done when stepping is enabled.

Reply:
Looking a little further down, it looks like this helper is only called when `self.step_state == StepState::Next` or `StepState::Rest` in the `match`.  Does that do what you're asking for here?

Reply:
Ah yeah, that's fine. It was hard to read on Github.